aws-datastore/src/main/java/com/amplifyframework/datastore/syncengine/ReachabilityMonitor.kt

```diff
@@ -91,34 +87,73 @@ private class ReachabilityMonitorImpl constructor(val schedulerProvider: Schedul
+        private var currentNetwork: Network? = null
+        private var currentCapabilities: NetworkCapabilities? = null
+
+        override fun onAvailable(network: Network) {
+            currentNetwork = network
+            updateAndSend()
+        }
+
+        override fun onLost(network: Network) {
+            currentNetwork = null
+            currentCapabilities = null
+            updateAndSend()
+        }
+
+        override fun onUnavailable() {
+            currentNetwork = null
+            currentCapabilities = null
+            updateAndSend()
+        }
+
+        override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
+            currentNetwork = network
+            currentCapabilities = networkCapabilities
+            updateAndSend()
+        }
+
+        private fun updateAndSend() {
+            emitter.onNext(NetworkCapabilitiesUtil.isInternetReachable(currentCapabilities))
+        }
+    }
 }
 
 /**
  * This interface puts an abstraction layer over ConnectivityManager. Since ConnectivityManager
  * is a concrete class created within context.getSystemService() it can't be overridden with a test
  * implementation, so this interface works around that issue.
  */
-public interface ConnectivityProvider {
+interface ConnectivityProvider {
     val hasActiveNetwork: Boolean
     fun registerDefaultNetworkCallback(context: Context, callback: NetworkCallback)
 }
 
 private class DefaultConnectivityProvider : ConnectivityProvider {
-
     private var connectivityManager: ConnectivityManager? = null
 
     override val hasActiveNetwork: Boolean
-        get() = connectivityManager?.let { it.activeNetwork != null }
-            ?: run {
-                throw DataStoreException(
-                    "ReachabilityMonitor has not been configured.",
-                    "Call ReachabilityMonitor.configure() before calling ReachabilityMonitor.getObservable()"
-                )
+        get() = connectivityManager?.let { manager ->
+            // the same logic as https://github.com/aws-amplify/amplify-android/blob/main/aws-storage-s3/src/main/java/com/amplifyframework/storage/s3/transfer/worker/BaseTransferWorker.kt#L176
+            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+                val networkCapabilities: NetworkCapabilities? = NetworkCapabilitiesUtil.getNetworkCapabilities(manager)
+                NetworkCapabilitiesUtil.isInternetReachable(networkCapabilities)
+            } else {
+                val activeNetworkInfo = manager.activeNetworkInfo
+                activeNetworkInfo != null && activeNetworkInfo.isConnected
             }
+        } ?: run {
+            throw DataStoreException(
+                "ReachabilityMonitor has not been configured.",
+                "Call ReachabilityMonitor.configure() before calling ReachabilityMonitor.getObservable()"
+            )
+        }
```

Comment:
Don't need the run block
```suggestion
        } ?: throw DataStoreException(
              "ReachabilityMonitor has not been configured.",
              "Call ReachabilityMonitor.configure() before calling ReachabilityMonitor.getObservable()"
          )
```